iml-agent/src/action_plugins/high_availability/corosync_conf.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::agent_error::ImlAgentError;
+use iml_cmd::{CheckedCommandExt, Command};
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::{collections::HashMap, convert::TryFrom, io};
+use tokio::{fs::OpenOptions, io::AsyncWriteExt};
+
+static COROSYNC_CONF: &str = "/etc/corosync/corosync.conf";
+
+/// Reads the corosync configuration.
+async fn read_corosync_conf() -> Result<Vec<u8>, io::Error> {
+    iml_fs::read_file_to_end(COROSYNC_CONF).await
+}
+
+/// Truncates and writes a new configuration.
+async fn write_corosync_conf(xs: &[u8]) -> Result<(), ImlAgentError> {
+    OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .open(COROSYNC_CONF)
+        .await?
+        .write_all(xs)
+        .await?;
+
+    Ok(())
+}
+
+/// Updates the config_version. This is useful so nodes can tell which conf is newer.
+fn update_config_version(
+    s: impl Iterator<Item = String>,
+    next_version: u64,
+) -> impl Iterator<Item = String> {
+    s.filter(|x| !x.trim().starts_with("config_version:"))
+        .map(move |x| {
+            if x.trim().starts_with("version:") {
+                format!("{}\n    config_version: {}", x, next_version)
+            } else {
+                x
+            }
+        })
+}
+
+/// Updates the mcastport for all interfaces in the corosync.conf.
+fn update_mcast_port(
+    s: impl Iterator<Item = String>,
+    new_port: u16,
+) -> impl Iterator<Item = String> {
+    lazy_static! {
+        static ref RE: Regex = Regex::new(r"(\s*mcastport:\s*)\d+").unwrap();
+    }
+
+    s.map(move |x| {
+        if x.trim().starts_with("mcastport:") {
+            RE.replace(&x, |caps: &regex::Captures| {
+                format!("{}{}", &caps[1], new_port)
+            })
+            .to_string()
+        } else {
+            x
+        }
+    })
+}
```

Comment:
This assumes mcastport is specified in config file

Reply:
Where else would it be specified?

If the `mcastport` is missing from the file, the regex replace is never run and this fn is effectively a noop.

Reply:
That's what I mean.  If it's missing in the file, it would show in `corosync-cmapctl totem.interface` output.  Defaults to port 5405.  This will fail silently.  Sort of an odd corner case, but if we're managing random corosync configs...

Reply:
I see... I'll fix in a followup